hdl21/flatten.py

```diff
@@ -0,0 +1,168 @@
+"""
+Flatten a module by moving all nested instances, ports and signals to the root level.
+
+See function `flatten()` for details.
+"""
+
+import copy
+from dataclasses import dataclass, field, replace
+from typing import Dict, Generator, Tuple, Union
+
+import hdl21 as h
+
+
+def _walk_conns(conns, parents=tuple()):
+    for key, val in conns.items():
+        if isinstance(val, dict):
+            yield from _walk_conns(val, parents + (key,))
+        else:
+            yield parents + (key,), val
+
+
+def _flat_conns(conns):
+    return {":".join(reversed(path)): value.name for path, value in _walk_conns(conns)}
+
+
+@dataclass
+class _PrimitiveNode:
+    inst: h.Instance
+    path: Tuple[h.Module, ...] = tuple()
+    conns: Dict[str, h.Signal] = field(default_factory=dict)  # type: ignore
+
+    def __str__(self):
+        name = self.make_name()
+        path = [p.name for p in self.path]
+        conns = _flat_conns(self.conns)
+        return str(dict(name=name, path=path, conns=conns))
+
+    def make_name(self):
+        return ":".join([p.name or "_" for p in self.path])
+
+
+def walk(
+    m: h.Module,
+    parents=tuple(),
+    conns=None,
+) -> Generator[_PrimitiveNode, None, None]:
+    if not conns:
+        conns = {**m.signals, **m.ports}
+    for inst in m.instances.values():
+        new_conns = {}
+        new_parents = parents + (inst,)
+        for src_port_name, sig in inst.conns.items():
+            if isinstance(sig, h.Signal):
+                key = sig.name
+            else:
+                raise ValueError(f"unexpected signal type: {type(sig)}")
+
+            new_sig_name = ":".join([p.name for p in parents] + [key])
+            if key in conns:
+                target_sig = conns[key]
+            elif key in m.signals:
+                target_sig = replace(m.signals[key], name=new_sig_name)
+            elif key in m.ports:
+                target_sig = replace(m.ports[key], name=new_sig_name)
+            else:
+                raise ValueError(f"signal {key} not found")
+            new_conns[src_port_name] = target_sig
+
+        if isinstance(inst.of, h.PrimitiveCall):
+            yield _PrimitiveNode(inst, new_parents, new_conns)
+        else:
+            yield from walk(inst.of, new_parents, new_conns)
+
+
+def _find_signal_or_port(m: h.Module, name: str) -> h.Signal:
+    """Find a signal or port by name"""
+
+    port = m.ports.get(name, None)
+    if port is not None:
+        return port
+
+    sig = m.signals.get(name, None)
+    if sig is not None:
+        return sig
+
+    raise ValueError(f"Signal {name} not found in module {m.name}")
+
+
+def is_flat(m: Union[h.Instance, h.Instantiable]) -> bool:
+    if isinstance(m, h.Instance):
+        return is_flat(m.of)
+    elif isinstance(m, (h.PrimitiveCall, h.GeneratorCall, h.ExternalModuleCall)):
```

Comment:
A `GeneratorCall` is flat? 
Why not flatten its `Module`-result?

Reply:
hm... this decision is kind arbitrary because I don't use `GeneratorCall` so it becomes a case I don't handle. Any suggestion how it should be handled here? Return `True` if its generated module is flat?